DBZ-6310 Document JDBC sink connector (#4433)
debezium/debezium: documentation/modules/ROOT/pages/connectors/jdbc.adoc

This exception occurs because some dialects map a `STRING` data type mapping to a variable length string data type such as `TEXT` or `CLOB`, and these dialects do not allow unbounded length primary key columns.
To avoid this problem, the following must be set:

* Do not set `schema.evolution` to `basic`.
* Create the database table and primary key mappings in advance.
====

[[jdbc-delete-mode]]
=== Delete mode

The {prodname} JDBC sink connector can delete rows in the destination database when a `DELETE` or _tombstone_ event is consumed.
By default, the JDBC sink connector does not enable delete mode, therefore, it must be enabled explicitly in the connector configuration if you wish to support removing rows.

Deletes can be enabled by setting `delete.enabled=true` in the connector configuration, but this mode requires that the `primary.key.mode` be set to a value that is not `none`.
This is because deletes are executed based on the primary key mapping, and therefore is a destination table has no primary key mapping, rows cannot be deleted by the connector.

[[jdbc-idempotent-writes]]
=== Idempotent writes

> The {prodname} JDBC sink connector supports idempotent writes, allowing the same records to be replayed repeatedly and the final database state remains consistent.

Reviewer: ```suggestion
The {prodname} JDBC sink connector supports idempotent writes, so that even if the same records are applied multiple times, the final database state is not changed after the first application.
```
Ok, I thnk I get it. Does this ^ work?

When an event is received from a topic and the record schema structure differs from the destination table , the {prodname} JDBC sink connector will use the event's key and/or record schema structure to resolve which columns are new that need to be added to the database table.
When schema evolution is enabled, an `ALTER TABLE` SQL statement will be prepared and executed by the connector before applying the DML event to the destination table.
The changing of column data types, dropping of columns, and adjustments to primary keys can be considered dangerous, and therefore none of these operations are performed by the connector.

Whether a column is `NULL` or `NOT NULL` is derived directly from the optionality defined in the field's schema, and default values are also resolved from the field schema as well.
If the connector attempts to create a table with an undesired nullability setting or default value, the table will either need to be created manually ahead of time or the schema of the associated field adjusted prior to be processing by the sink connector.

> This can be accomplished by introducing a custom single message transformation that adjusts these in the pipeline or modifying the column state in the source database.

Author: What I am trying to convey is that if the source connector emits a field with an undesirable schema type, the user can adjust this by adding a single message transformation (SMT) to the pipeline either on the source before going into Kafka or the sink as the JDBC sink connector reads the data from Kafka, they can add a `CustomConverter` on the source connector, or they can alter the default value or nullability at the table level in the source database.

By enabling the column or data type propagation in the source connector, the {prodname} JDBC sink connector can use this information to refine the data type mapping process and will create the column as `TINYINT(1)`.

[NOTE]
====
Typically, column / data type propagation will have a much larger influence when both the source and sink databases are the same.
We are continually looking at ways to improve this mapping across heterogeneous databases and the current type system allows us to continue to refine these mappings based on feedback.
If you find a mapping could be improved, please let us know.
====

[[jdbc-deployment]]
== Deployment

To deploy a {prodname} JDBC connector, you install the {prodname} JDBC connector archive, configure the connector, and start the connector by adding its configuration to Kafka Connect.

.Prerequisites
* link:https://zookeeper.apache.org/[Apache ZooKeeper], link:http://kafka.apache.org/[Apache Kafka], and link:{link-kafka-docs}.html#connect[Kafka Connect] are installed.

> * Destiniation database is installed and configured to accept JDBC connections.

Reviewer 2: ```suggestion
* Destination database is installed and configured to accept JDBC connections.
```

.Procedure

. Download the {prodname} https://repo1.maven.org/maven2/io/debezium/debezium-connector-jdbc/{debezium-version}/debezium-connector-oracle-{debezium-version}-plugin.tar.gz[JDBC connector plug-in archive].
. Extract the files into your Kafka Connect environment.
. Optionally download the JDBC driver from Maven Central and extract the downloaded driver file to the directory that contains the JDBC sink connector JAR file.
+
NOTE: Drivers for Oracle and Db2 are not automatically shipped with the JDBC sink connector and must be manually installed.

> . Add the directory with the JAR files to {link-kafka-docs}/#connectconfigs[Kafka Connect's `plugin.path`].

Reviewer 2: I am not sure about this wording. We usually don't add plug-in directory to the `plugin.path` but the directory that contains plugin directory.

Author: I'll adjust this wording, but this was taken from the Oracle documentation, so @roldanbob we may need to address this in our other connector docs as well and unify the language.

|===

[[jdbc-connector-properties-extendable]]
.Extendable properties
[cols="30%a,25%a,45%a"]
|===
|Property |Default |Description

|[[jdbc-property-column-naming-strategy]]<<jdbc-property-column-naming-strategy, `+column.naming.strategy+`>>
|`i.d.c.j.n.DefaultColumnNamingStrategy`
|Specifies the fully-qualified class name of a `ColumnNamingStrategy` implementation that the connector will use to resolve column names from event field names. +
 +
The default behavior is to simply use the field name as the column name.
|[[jdbc-property-table-naming-strategy]]<<jdbc-property-table-naming-strategy, `+table.naming.strategy+`>>
|`i.d.c.j.n.DefaultTableNamingStrategy`
|Specifies the fully-qualified class name of a `TableNamingStraetgy` implementation that the connector users to resolve table names from incoming event topic names. +
 +
The default behavior is to: +

* Replace the `${topic}` placeholder in the `table.name.format` configuration property with the event's topic.
* Sanitize the table name by replacing dots (`.`) with underscores (`_`).

|===

[[jdbc-faq]]
== Frequently asked questions

* *Is the* `ExtractNewRecordState` *single message transformation required?* +

> No, that is actual one of the differentiating factors of the {prodname} JDBC connector from its competitors.

Reviewer 2: Do we want to use term `competitor`? Maybe `other solution` would be better?

|`enum`
|Types.VARCHAR
|n/a

|io.debezium.data.Json
|`json`
|`json`
|n/a

|io.debezium.data.EnumSet
|`set`
|n/a
|n/a

|io.debezium.time.Year
|`year(n)`
|n/a
|n/a

|io.debezium.time.MicroDuration
|n/a
|`interval`
|n/a

|io.debezium.data.Ltree
|n/a
|`ltree`
|n/a

|io.debezium.data.Uuid
|n/a
|`uuid`
|n/a

|io.debezium.data.Xml
|n/a
|`xml`
|`xml`

|===

In addition to the primitive and logical mappings above, if the source of the change events is a {prodname} source connector, the resolution of the column type, its length, precision, and scale can be further influenced by enabling column or data type propagation.
This propagation requires that the source connector be configured with `column.propagate.source.type` or `datatype.propagate.source.type` and the {prodname} JDBC sink connector will use these values with higher precedence.

For example, let's say the following field schema is included in a change event:

.{prodname} change event field schema with column/data type propagation enabled
[source,json]
----
{
  "schema": {
    "type": "INT8",
    "parameters": {
      "__debezium.source.column.type": "TINYINT",
      "__debezium.source.column.length": "1"
    }
  }
}
----

> In the above example, without the schema parameters, the {prodname} JDBC sink connector would map this field to a column type of `Types.SMALLINT`, which will have varying logical database types depending on the dialect but for MySQL this would be a `TINYINT` column type without any length.

Author: I think _no specified length_ makes logical sense.  I'll apply that wording instead.

These types of schema changes should be handled manually.

* *A column's type is resolved to a type that is not desired, how can the type be resolved to a different data type?* +
The {prodname} JDBC connector uses a sophisticated type system to resolve a column's data type.
You can refer to the xref:#jdbc-data-and-type-mappings[] section for details on how this type system resolves a specific field's schema definition to a JDBC type.
If you desire a different data type mapping, you should define the table manually so you explicitly control the desired column type you need.

* *How do you specify a prefix or a suffix to the table name without changing the Kafka topic name?* +
In order to add a prefix or a suffix to the destination table  name, simply adjust the xref:#jdbc-property-table-name-format[table.name.format] connector configuration property and prepend or append the desired prefixed or suffix.
For example, to prefix all table names with `jdbc_`, specify the `table.name.format` configuration property with a value of `jdbc_${topic}`.
If such a connector is subscribed to a topic called `orders`, the resulting table will be created as `jdbc_orders`.

> * *Even when identifier quoting is not enabled, some columns are automatically quoted, why?* +

Reviewer: ```suggestion
* *Why are some columns automatically quoted, even though identifier quoting is not enabled?* +
```

> There are situations where specific column or table names may still be explicitly quoted even when `quote.identifiers` is not enabled.

Reviewer: ```suggestion
In some situations, specific column or table names might be explicitly quoted, even when `quote.identifiers` is not enabled.
```

This is often necessary when the column or table name starts with or uses a specific convention that would otherwise be considered illegal syntax.

> For example, when setting the xref:#jdbc-property-primary-key-mode[primary.key.mode] to `kafka`, certain databases will not allow a column's name to begin with an underscore unless the column's name is quoted.

Reviewer: ```suggestion
For example, when the xref:jdbc-property-primary-key-mode[primary.key.mode] is set to `kafka`, some databases only permit column names to begin with an underscore if the column's name is quoted.
```

> This behavior is dialect specific and varies depending on the database used.

Reviewer: ```suggestion
Quoting behavior is dialect-specific, and varies among different types of database.
```
